### ISSUE TITLE
Go modules: update source code after major updates

### ISSUE THREAD

**Issue by rarkins** [renovatebot/renovate#3348]
**What would you like Renovate to be able to do?**

Update Go's import references after major updates

**Describe the solution you'd like**

Whenever doing a major update in `go.mod`, Renovate should find and update all "old" references from within source code, e.g.

![image](https://user-images.githubusercontent.com/6311784/54102827-38fbed80-43ca-11e9-9299-c828a11e946f.png)

I'm not sure yet about exactly what logic Renovate should use or if it's as simple as a global grep/search/replace through `*.go` files.

**Describe alternatives you've considered**

Otherwise users need to add these changes themselves to a PR.

**Additional context**

This is an example PR from Renovate today: https://github.com/hairyhenderson/github-responder/pull/82

Here is a manual/human PR that replaced it: https://github.com/hairyhenderson/github-responder/pull/90 (Cc @hairyhenderson)

Something also strange is that the replacement PR has vendored modules updated while Renovate's did not. I'm suspecting it could be related to Renovate's failure to update the references in source code, but not 100% sure.

**Comment by rarkins** [comment]
I checked this manually and verified that `go mod vendor` indeed works as expected only once the import statements are updated. Example command I used: `find *.go -type f -print0 | xargs -0 sed -i '' 's/"github.com\/google\/go-github\/v20/"github.com\/google\/go-github\/v24/g'`

**Comment by hairyhenderson** [comment]
> I'm not sure yet about exactly what logic Renovate should use or if it's as simple as a global grep/search/replace through `*.go` files.

That's probably enough IMO - if there are incompatibilities, the code won't compile, and that should be obvious through CI.

**Comment by rarkins** [comment]
Thanks, I think so too. So once this is implemented you would see the exact same PR that you generated by hand, including vendored modules. In cases where the breaking changes break your code, you hopefully see that either by a build error or tests breakage, but either way you should know to check major updates carefully before merging :)

**Comment by hairyhenderson** [comment]
Thanks @rarkins! 😬

**Comment by R-omk** [comment]
Here is a tool that does exactly what you need
https://github.com/marwan-at-work/mod

**Comment by 56KBs** [comment]
I have looked at a potential implementation of this functionality - Focused around using the `marwan-at-work/mod` tool within renovate.

The implementation itself appears to be fairly straightforward when making us of the `mod` tool, requiring changes similar to the following within `lib/manager/gomod/artifacts.ts`:

```go
    // Update import paths for major updates
    const isImportPathUpdateRequired = !depName.startsWith("gopkg.in") && config.updateType === 'major' && _updatedDeps.length > 0;
    if (isImportPathUpdateRequired) {
      const modCmd = 'mod';
      args = `upgrade --mod-name=${_updatedDeps[0]} -t=${config.newMajor}`;
      logger.debug({ modCmd, args }, 'marwen-at-work/mod command included');
      execCommands.push(`${modCmd} ${args}`);
    }
```
```go
    // Include all the .go file import changes
    if (isImportPathUpdateRequired) {
      for (const f of status.modified.concat(status.not_added)) {
        if (f.endsWith('.go')) {
          res.push({
            file: {
              name: f,
              contents: await readLocalFile(f),
            },
          });
        }
      }
    }
```

Please note: I'm aware there are potential issues in this code - This is purely for a proof of concept

In terms of installation of the `mod` tool, this can be installed via Go's existing tooling using `go install github.com/marwan-at-work/mod/cmd/mod@v0.4.1` meaning it could be installed during runtime of renovate, if so desired.

The advantages of using this tool is that is using the core golang tooling in order to identify and update the imports for the provided package you're working on, meaning it should be less error prone. Additionally, the creator of the tool is a GitHub employee and one of the largest contributors to the Go [athens proxy](https://github.com/gomods/athens).

Alternative solutions could exist, making use of native Javascript, rather than external tooling:

1. Recursive find and replace of all `.go` files (excluding the vendor directory)
1. Implement file parsing for all `.go` files, identifying `import` statements and updating them, similar to how `lib/manager/gomod/extract.ts` works on `go.mod files`

As a side note, this appears to be a duplicate of #3627.

**Comment by rarkins** [comment]
> In terms of installation of the `mod` tool, this can be installed via Go's existing tooling using `go install github.com/marwan-at-work/mod/cmd/mod@v0.4.1` meaning it could be installed during runtime of renovate, if so desired.

Would this modify the existing `go.mod`? Does `go` have the concept of installing something "globally" into $PATH like an `npm install --global` does?

I think it's a good idea to go with this tool, for the reasons you mention.

**Comment by 56KBs** [comment]
> Would this modify the existing `go.mod`? Does `go` have the concept of installing something "globally" into $PATH like an `npm install --global` does?

No modifications will occur to a `go.mod` file when installing with a version suffix e.g `@latest` or `@v0.4.1`. Executables will be installed $GOBIN (Defaults to $GOPATH/bin or $HOME/go/bin if the $GOPATH is not set).

Assuming the $GOBIN directory is set to somewhere on $PATH it would effectively result in a global install similar to npm

**Comment by rarkins** [comment]
Great. We'd need to make sure our prebuilt containers are set up like that but sounds workable. In that case it seems like it can be a postUpdateOption or even default behaviour for major updates *if* were confident nobody will need to opt out
